[doc] MultiKueue (#1705)
kubernetes-sigs/kueue: site/content/en/docs/admission-check-controllers/multikueue.md

---
title: "MultiKueue Admission Check Controller"

> date: 2023-10-23

Reviewer: update

Author: done

weight: 2
description: >
  An admission check controller providing the manager side MultiKueue functionality.
---

> The MultiKueue Admission Check Controller is an Admission Check Controller designed to provide the manager side functionality for [MultiKueue](https://github.com/kubernetes-sigs/kueue/blob/main/keps/693-multikueue/README.md). 

Reviewer 2: ```suggestion
The MultiKueue Admission Check Controller is an Admission Check Controller designed to provide the manager side functionality for multi cluster job dispatching.
```
Let's avoid linking to the KEP, it's not meant for end users

> The controller is part of kueue. You can enable it by setting the `MultiKueue` feature gate. Check the [Installation](/docs/installation/#change-the-feature-gates-configuration) guide for details on feature gate configuration.

Reviewer 2: ```suggestion
You can enable it by setting the `MultiKueue` feature gate. Check the [Installation](/docs/installation/#change-the-feature-gates-configuration) guide for details on feature gate configuration.
```

{{% alert title="Warning" color="warning" %}}
MultiKueue is currently an alpha feature and disabled by default.
{{% /alert %}}

> The controller's main attributes are:

Reviewer: Maybe we should have something more user / admin - oriented. I'm thinking, what would be particularly useful:
1  (admin pov) demonstrate step-by-step how admin would setup the clusters (installing jobset etc. on the workers, but only CRDs), with example configs for the manager and worker clusters 
2. (user pov) explain how user can interact with the environment, explaining that users interacts only with the manager cluster, but should not expect the pods to be created there. If they are interested in debugging the job they need to figure out the worker the job is actually running on, connect to the cluster and see pods.

This could probably be follow ups for the admin / user tasks. WDYT @tenzen-y @alculquicondor @mwielgus ?

Reviewer 3: > This could probably be follow ups for the admin / user tasks. WDYT @tenzen-y @alculquicondor @mwielgus ?

That makes sense. 
Also, I think we should clarify the responsibilities of each role (user, admin, kueue-controller-manager), like kubeconfig, in the `task` section.

Reviewer 2: Indeed, the current paragraph contains implementation details.

We need to briefly start with the architecture (manager, workers) and when jobs are sent to the workers (first they need to pass quota checks in the manager, then the manager makes the worker clusters compete to run the job).

We should also mention that the manager cluster should have a quota roughly equal to the total quota among all clusters.

- Establish and maintain the connection with the worker clusters.
- Maintain the `Active` status of the Admission Checks controlled by `multikueue`

> - Crate and monitor remote objects (workloads or jobs) while keeping the local ones in sync.

Reviewer: ```suggestion
- Create and monitor remote objects (workloads or jobs) while keeping the local ones in sync.
```

Author: done

There is an ongoing effort to overcome these limitations by adding the possibility to disable the reconciliation of some jobs by the main `batch/Job` controller. Details in `kubernetes/enhancements` [KEP-4368](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-label-for-batch-jobs#readme).

### JobSet
Known Limitations:
- Since unsuspending a JobSet in the manager cluster will lead to its local execution and updating the status of a local JobSet could conflict with its main controller, only the JobSet CRDs should be installed in the manager cluster.
An approach similar to the one described for [`batch/Job`](#batchjob) is taken into account to overcome this. 

## Parameters

An AdmissionCheck controlled by `multikueue` should use a `kueue.x-k8s.io/v1alpha1` `MultiKueueConfig` parameters object, details of its structure can be found in [Kueue Alpha API reference section](/docs/reference/kueue-alpha.v1alpha1/#kueue-x-k8s-io-v1alpha1-MultiKueueConfig)

## Example

> ### Setup

Reviewer 2: Before giving an example, we also need to explain how the user needs to configure the clusters to have the same namespaces and clusterqueues.

{{< include "/examples/multikueue/multikueue-setup.yaml" "yaml" >}}

For the example provided, having the worker1 cluster kubeconfig stored in a file called `worker1.kubeconfig`, the 
`worker1-secret` secret ca by created with:

```bash

>  kubectl create secret generic worker1-secret -n kueue-system --from-file=kubeconfig=worker1.kubeconfig

Reviewer: Maybe this is another section / page, but I think it would be very useful to extend the example with :
- setup for the worker clusters
- running an example jobset - demonstrate how a user to monitor it, in particular figure out which cluster it is running on,  demonstrate

Author: It could be a multikue specific section in `Instalation`, but when we'll support "managed-by" , there will be no additional steps to do.

Strictly for how to create a secret out of a file ... I don't think it's very special .. we can also drop it from here.

Reviewer: > It could be a multikue specific section in Instalation

Something analogous to this page I'm thinking https://kueue.sigs.k8s.io/docs/tasks/administer_cluster_quotas/, but for MultiKueue.

I wouldn't  mix this in regular installation. I think MultiKueue deserves its own page in tasks for admins. Maybe from the installation we could reference the task. 

I think we should show example configurations for the manager and worker clusters. Describe what is the relationship between the quotas, at least briefly. It is far from obvious if the quotas should be the same or different.

---
title: "MultiKueue Admission Check Controller"
date: 2024-02-08
weight: 2
description: >

>   An admission check controller providing the manager side MultiKueue functionality.

Reviewer 2: ```suggestion
  An admission check controller for multi cluster job dispatching.
```

Author: done

---

> 

Author: done

> The MultiKueue Admission Check Controller is an Admission Check Controller designed to provide the manager side functionality for multi cluster job dispatching.

Reviewer 2: ```suggestion
MultiKueue is an [admission check](/docs/concepts/admission_check/) controller designed to provide the manager side functionality for multi cluster job dispatching.
```

Author: done

You can enable it by setting the `MultiKueue` feature gate. Check the [Installation](/docs/installation/#change-the-feature-gates-configuration) guide for details on feature gate configuration.

{{% alert title="Warning" color="warning" %}}
MultiKueue is currently an alpha feature and disabled by default.
{{% /alert %}}

## MultiKueue Overview

A MultiKueue setup is composed of a manager cluster and at least one worker cluster.

### Cluster Roles
#### Manager Cluster

> The manager's main attributes are:

Reviewer 2: ```suggestion
The manager's main responsibilities are:
```

Author: done

- Establish and maintain the connection with the worker clusters.
- Create and monitor remote objects (workloads or jobs) while keeping the local ones in sync.

The MultiKueue Admission Check Controller runs in the manager cluster and will also maintain the `Active` status of the Admission Checks controlled by `multikueue`.

> The ResourceQuota set on a ClusterQueue using MultiKueue controls how many jobs are subject for dispatching at a given point in time, ideally it should be less then the total quotas in the worker clusters since otherwise will cause the monitoring of workloads that don't have a chance to be admitted. 

Reviewer 2: ```suggestion
The ResourceQuota set on a ClusterQueue using MultiKueue controls how many jobs are subject for dispatching at a given point in time.
Ideally, the quota in the manager cluster should be equal to the total quotas in the worker clusters.
If significantly lower, the worker clusters will be under utilized.
If significantly higher, the manager will dispatch and monitor workloads in the worker clusters that don't have a chance to be admitted.
```

Author: done

> The ResourceQuota set on a ClusterQueue using MultiKueue controls how many jobs are subject for dispatching at a given point in time, ideally it should be less then the total quotas in the worker clusters since otherwise will cause the monitoring of workloads that don't have a chance to be admitted. 

Reviewer 2: flavorQuota is not a field name either, so maybe we should say "The quota set for the flavors of a ClusterQueue"

> The ResourceQuota set on a ClusterQueue using MultiKueue controls how many jobs are subject for dispatching at a given point in time, ideally it should be less then the total quotas in the worker clusters since otherwise will cause the monitoring of workloads that don't have a chance to be admitted. 

Reviewer 3: That makes sense.

> The ResourceQuota set on a ClusterQueue using MultiKueue controls how many jobs are subject for dispatching at a given point in time, ideally it should be less then the total quotas in the worker clusters since otherwise will cause the monitoring of workloads that don't have a chance to be admitted. 

Author: done

#### Worker Cluster

> The worker cluster acts like an usual standalone Kueue cluster in which the Kueue the workloads and jobs are managed by the MultiKueue Admission Check Controller running in the manager cluster.

Reviewer 2: ```suggestion
The worker cluster acts like a standalone Kueue cluster.
The workloads and jobs in the worker cluster are created or deleted by the
MultiKueue Admission Check Controller running in the manager cluster.
```

Author: done

### Job Flow

> If the ClusterQueue a job is assigned to in the manager cluster uses a MultiKueue AdmissionCheck it will be the subject of dispatching, as a result:

Reviewer 2: ```suggestion
For a job to be subject to multi cluster dispatching, you need to assign it to a ClusterQueue 
that uses a MultiKueue AdmissionCheck. The Multikueue system works as follows:
```

Author: done

> - When it's workload gets a QuotaReservation in the manager, a copy of that workload will be created in all the configured worker clusters.

Reviewer 2: ```suggestion
- When the job's Workload gets a QuotaReservation in the manager cluster, a copy of that workload will be created in all the configured worker clusters.
```

Author: done

> - When one of the remote workloads is admitted in a worker cluster:

Reviewer 2: ```suggestion
- When one of the worker clusters admits the remote workload sent to it:
```

Author: done

>   - All the other remote workloads are removed.

Reviewer 2: ```suggestion
  - The manager removes all the other remote workloads.
```

Author: done

>   - A copy of the job is created in the selected worker cluster having the remote workload set as its prebuilt-workload.

Reviewer 2: ```suggestion
  - The manager creates a copy of the job in the selected worker cluster, configured to use the quota reserved by the admitted Workload.
```

Author: done

>   - A copy of the job is created in the selected worker cluster having the remote workload set as its prebuilt-workload.

Reviewer 3: Can we mention the prebuild annotation here?

Author: done

>   - The AdmissionCheckState of the local workload is set as `Ready` resulting in its Admission in the manager cluster.

Reviewer 2: ```suggestion
  - The manager sets the AdmissionCheckState of the local workload to `Ready`, resulting in its Admission in the manager cluster.
```

Reviewer 2: Although we could remove this line. It's an implementation detail.

Author: done

> - The remote objects , workload and job, are monitored and the changes in their status are synced back in the manager objects.

Reviewer 2: ```suggestion
- The manager monitors the remote objects, workload and job, and syncs any changes in their status into the local objects.
```

Author: done

- When the remote workload is marked as `Finished`:

>   - A last sync is done for the objects status.

Reviewer 2: ```suggestion
  - The manager does a last sync for the objects status.
```

Author: done

>   - The remote objects are removed from the worker cluster.

Reviewer 2: ```suggestion
  - The manager remotes the objects from the worker cluster.
```

Author: done

> ## Supported jobs

Reviewer 2: and empty lines after titles

Author: done

### batch/Job
Known Limitations:
- Since unsuspending a Job in the manager cluster will lead to its local execution, the AdmissionCheckStates are kept `Pending` during the remote job execution.

> - Since updating the status of a local Job could conflict with the Job's main controller, the Job status is not synced during the job execution, the final status of the remote Job is copied when the remote workload is marked as `Finished`.

Reviewer 2: ```suggestion
- Since updating the status of a local Job could conflict with the Kubernetes Job controller, the manager
  does not sync the Job status during the job execution.
  The manager copies the final status of the remote Job when the remote workload is marked as `Finished`.
```

Author: done

> There is an ongoing effort to overcome these limitations by adding the possibility to disable the reconciliation of some jobs by the main `batch/Job` controller. Details in `kubernetes/enhancements` [KEP-4368](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-label-for-batch-jobs#readme).

Reviewer 2: ```suggestion
There is an ongoing effort to overcome these limitations by adding the possibility to disable the reconciliation of some jobs by the Kubernetes `batch/Job` controller. Details in `kubernetes/enhancements` [KEP-4368](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-label-for-batch-jobs#readme).
```

Author: done

### JobSet
Known Limitations:

> - Since unsuspending a JobSet in the manager cluster will lead to its local execution and updating the status of a local JobSet could conflict with its main controller, only the JobSet CRDs should be installed in the manager cluster.

Reviewer 2: ```suggestion
- Since unsuspending a JobSet in the manager cluster will lead to its local execution and updating the status of a local JobSet could conflict with its main controller, you should only install the JobSet CRDs, but not the controller.
```

Author: done

An approach similar to the one described for [`batch/Job`](#batchjob) is taken into account to overcome this. 

> 

Reviewer 3: I think that we can say that deploying jobset-controller to the manager cluster is possible when we use the jobset with the main branch, right?

Author: We don't yet support it in MultiKueue ACC.

## Parameters

An AdmissionCheck controlled by `multikueue` should use a `kueue.x-k8s.io/v1alpha1` `MultiKueueConfig` parameters object, details of its structure can be found in [Kueue Alpha API reference section](/docs/reference/kueue-alpha.v1alpha1/#kueue-x-k8s-io-v1alpha1-MultiKueueConfig)

## Example

### Setup

#### Worker Cluster

> When a workload is dispatched from the manager cluster to a worker cluster it is expected that its namespace and LocalQueue to be already created, hence the worker cluster configuration should mirror the one of the manager cluster in terms of namespaces and LocalQueues.

Reviewer 2: ```suggestion
When MultiKueue dispatches a workload from the manager cluster to a worker cluster, it expects that
the job's namespace and LocalQueue also exist in the worker cluster.
In other words, you should ensure that the worker cluster configuration mirrors the one of the manager cluster in terms of namespaces and LocalQueues.
```

Author: done

#### Manager Cluster

{{< include "/examples/multikueue/multikueue-setup.yaml" "yaml" >}}

> For the example provided, having the worker1 cluster kubeconfig stored in a file called `worker1.kubeconfig`, the 
> `worker1-secret` secret ca by created with:

Reviewer 2: ```suggestion
For the example provided, having the worker1 cluster kubeconfig stored in a file called `worker1.kubeconfig`,
you can create the 
`worker1-secret` secret by running the following command:
```

Author: done
